### PR TITLE
[FLOPs] Fix attention FLOPs estimate

### PR DESCRIPTION
**Pull request by shuhuayu**
Fix the FLOPs estimate in non-parameterized attention matmuls. This part estimate of FLOPs is customized for each model due to different Attention mechanisms.

### CODE REVIEW

File: torchtitan/models/utils.py [hunk 1]
```diff
@@ -391,11 +391,11 @@ def get_dense_model_nparams_and_flops(
     return nparams, num_flops_per_token
 
 
-def get_moe_model_nparams_and_flops(
+def get_moe_model_nparams_and_para_linear_flops(
     model_args: BaseModelArgs, model: nn.Module, seq_len: int
 ) -> tuple[int, float]:
     """
-    Calculate nparams and nflops for MoE model
+    Calculate nparams and parameterized linear nflops (excluding non-parameter attention matmuls) for MoE model
```
Comment (wwwjn) [inline torchtitan/models/utils.py]: What does "non-parameter" attention matmul mean? The name is a little bit hard to understand,

Comment (shuhuayu) [reply torchtitan/models/utils.py]: By non-parameter attention matmul, I mean the portion of FLOPs in the attention block that does not involve parameterized weight matrices but primarily depends on the sequence length.

File: torchtitan/models/utils.py [hunk 2]
```diff
@@ -429,20 +429,12 @@ def get_moe_model_nparams_and_flops(
...
+    num_para_linear_flops_per_token = 6 * (
+        nparams_dense - nparams_embedding + nparams_sparse_active
     )
 
     # If weight tying is enabled, subtract embedding parameters from total count
     if hasattr(model_args, "enable_weight_tying") and model_args.enable_weight_tying:
         nparams = nparams - nparams_embedding
 
-    return nparams, num_flops_per_token
+    return nparams, num_para_linear_flops_per_token
```
Comment (tianyu-l) [inline torchtitan/models/utils.py]: For simplicity, do you think we can name this `num_linear_flops_per_token`? We could name the other part `num_attention_flops_per_token`.

File: torchtitan/models/deepseek_v3/model/args.py [hunk 3]
```diff
@@ -113,4 +113,18 @@ def update_from_config(self, job_config: JobConfig, **kwargs) -> None:
...
-        return get_moe_model_nparams_and_flops(self, model, seq_len)
+        (
+            nparams,
+            num_para_linear_flops_per_token,
+        ) = get_moe_model_nparams_and_para_linear_flops(self, model, seq_len)
+        num_nonpara_attn_flops_per_token = (
+            6
+            * self.n_layers
+            * self.n_heads
+            * (self.qk_nope_head_dim + self.qk_rope_head_dim + self.v_head_dim)
```
Comment (tianyu-l) [inline torchtitan/models/deepseek_v3/model/args.py]: I thought more.

If it's only the `head_dim` varying from model to model, maybe we can just send `head_dim` as an arg to `get_nparams_and_flops` just like `seq_len`? It sound cleaner to me this way. WDYT?

File: torchtitan/models/utils.py [hunk 4]
```diff
@@ -391,11 +391,11 @@ def get_dense_model_nparams_and_flops(
     return nparams, num_flops_per_token
```
Comment (tianyu-l) [inline torchtitan/models/utils.py:405]: For consistency, maybe we should refactor dense part as well?

File: torchtitan/models/utils.py (hunk 2 above)
Comment (wwwjn) [reply torchtitan/models/utils.py]: I agree, this name is easier to understand

File: torchtitan/models/deepseek_v3/model/args.py (hunk 3 above)
Comment (shuhuayu) [reply torchtitan/models/deepseek_v3/model/args.py]: I’m thinking about it this way: if the function `get_nparams_and_flops` is meant for users to quickly check FLOPs counts, they shouldn’t need to deal with the details of how `head_dim` values are computed, especially in cases like DeepSeek V3, where `qk_head_dim` is inferred from the attention mechanism and differs from `v_head_dim`. I’d prefer to hide these internal details from the user and let them simply provide the `seq_len` they use in their training configuration.

Comment (tianyu-l) [reply torchtitan/models/deepseek_v3/model/args.py]: No we don't change the signature of this function. Instead, we should change the signature of the function in `utils.py`. The function here is model specific and should internally code what the `head_dim` is.

Besides, the`head_dim` can be default to None and inferred from `model_args.dim // model_args.n_heads`.

File: torchtitan/models/utils.py (hunk 2 above)
Comment (shuhuayu) [reply torchtitan/models/utils.py]: Thanks for the suggestion to make it simple! I am thinking using `num_para_flops_per_token` to denote the FLOPs from matmuls involving parametrized weight matrices, and `num_nonpara_flops_per_token` for non-parameterized  matmuls (not involving parameterized weight matrices) in computing attention scores and value aggregation (with complexity O(`seq_len * head_dim`). Other non-parametrized flops per token including softmax, norms, and RoPE are neglected since they are only linear in  `head_dim`.

File: torchtitan/models/deepseek_v3/model/args.py (hunk 3 above)
Comment (shuhuayu) [reply torchtitan/models/deepseek_v3/model/args.py]: Got it! `get_nparams_and_flops` in `args.py` still has the current signature, and functions like`get_moe_model_nparams_and_para_flops` in `utils.py` will take both `seq_len` and `head_dim`. That makes great sense.

File: torchtitan/experiments/gpt_oss/__init__.py [hunk 5]
```diff
@@ -8,7 +8,7 @@
 from torchtitan.components.lr_scheduler import build_lr_schedulers
 from torchtitan.components.optimizer import build_optimizers_with_moe_load_balancing
 from torchtitan.components.tokenizer import build_hf_tokenizer
-from torchtitan.datasets.hf_datasets import build_hf_dataloader
+from torchtitan.hf_datasets.text_datasets import build_text_dataloader
```
Comment (wwwjn) [inline torchtitan/experiments/gpt_oss/__init__.py:11]: Thanks for helping fixing this!

File: torchtitan/models/utils.py [hunk 6]
```diff
@@ -361,19 +361,22 @@ def _concatenate_expert_weights(
 
 
 def get_dense_model_nparams_and_flops(
-    model_args: BaseModelArgs, model: nn.Module, seq_len: int
+    model_args: BaseModelArgs, model: nn.Module, head_dims: int, seq_len: int
 ) -> tuple[int, float]:
+    """
+    head_dims: the sum of qk and v head dims
```
Comment (wwwjn) [inline torchtitan/models/utils.py]: If adding docstring here, could also add all parameters?

File: torchtitan/models/utils.py [hunk 7]
```diff
@@ -392,10 +395,14 @@ def get_dense_model_nparams_and_flops(
...
-    model_args: BaseModelArgs, model: nn.Module, seq_len: int
+    model_args: BaseModelArgs,
+    model: nn.Module,
+    head_dims: int,
+    seq_len: int,
 ) -> tuple[int, float]:
     """
-    Calculate nparams and nflops for MoE model
+    Calculate nparams and parameterized linear nflops (excluding non-parameter attention matmuls) for MoE model
+    head_dims: the sum of qk and v head dims
```
Comment (wwwjn) [inline torchtitan/models/utils.py]: Same here, for consistency

File: torchtitan/models/llama3/model/args.py [hunk 8]
```diff
@@ -63,4 +63,6 @@ def update_from_config(self, job_config: JobConfig, **kwargs) -> None:
     def get_nparams_and_flops(
         self, model: nn.Module, seq_len: int
     ) -> tuple[int, float]:
-        return get_dense_model_nparams_and_flops(self, model, seq_len)
+        return get_dense_model_nparams_and_flops(
+            self, model, 2 * self.dim // self.n_heads, seq_len
```
Comment (wwwjn) [inline torchtitan/models/llama3/model/args.py]: Do we need parenthesis here to make the calculation correct?
```suggestion
            self, model, 2 * (self.dim // self.n_heads), seq_len
```

Comment (shuhuayu) [reply torchtitan/models/llama3/model/args.py]: Yes, in general they are different and should be updated. In `llama3` models, `dim` is always divisible by `n_heads`, so it should be correct.

File: torchtitan/models/utils.py [hunk 9]
```diff
@@ -361,19 +361,31 @@ def _concatenate_expert_weights(
...
 
-    l, h, q, t = (
+    l, h, qv, t = (
         model_args.n_layers,
         model_args.n_heads,
-        model_args.dim // model_args.n_heads,
+        head_dims,
         seq_len,
     )
     # Reasoning behind the factor of 12 for the self-attention part of the formula:
```
Comment (tianyu-l) [inline torchtitan/models/utils.py]: It seems we need to update the comment as well. Now there is no "12". We can be more specific about what `6 * l * h * qv * t` is about, and what is not taken into consideration.

File: torchtitan/models/utils.py [hunk 10]
```diff
@@ -361,19 +361,31 @@ def _concatenate_expert_weights(
...
         if isinstance(m, nn.Embedding)
     )
 
-    l, h, q, t = (
+    l, h, qv, t = (
         model_args.n_layers,
         model_args.n_heads,
-        model_args.dim // model_args.n_heads,
+        head_dims,
         seq_len,
```
Comment (tianyu-l) [inline torchtitan/models/utils.py]: Do you think it's better if we set `n_layers` and `n_heads` as args to this function? I feel it might be more clear as a pure util function. E.g. a model can define number of layers using `model_args.num_layers` or `model_args.L` but still be able to use this util function.

We can even consider putting it to tools/utils.py instead of model/utils.py.

File: torchtitan/models/utils.py [hunk 11]
```diff
@@ -361,19 +361,31 @@ def _concatenate_expert_weights(
...
+    """
     nparams = sum(p.numel() for p in model.parameters())
     nparams_embedding = sum(
         sum(p.numel() for p in m.parameters())
         for m in model.children()
         if isinstance(m, nn.Embedding)
     )
 
-    l, h, q, t = (
+    l, h, qv, t = (
```
Comment (tianyu-l) [inline torchtitan/models/utils.py]: It might be better to be more specific on the naming `n_layers` `n_heads`, etc. especially for `qv` which looks a bit confusing.

File: torchtitan/models/utils.py (hunk 10 above)
Comment (shuhuayu) [reply torchtitan/models/utils.py]: I agree. Do you think we can also move the following part to `args.py` so `get_moe_model_nparams_and_flops` does not need to take `model_args` as input, and thus become `model_args` agnostic? 
https://github.com/pytorch/torchtitan/blob/06ec49503ea9e1dc228aa8aa1b19d7c7b3043e3c/torchtitan/models/utils.py#L445-L446

For your last point, I think these two functions are still more model relevant than those functions currently in `tools/utils.py`.

File: torchtitan/models/utils.py (hunk 9 above)
Comment (shuhuayu) [reply torchtitan/models/utils.py]: Thanks for the catch, we should update the reasoning part.

File: torchtitan/models/utils.py (hunk 10 above)
Comment (tianyu-l) [reply torchtitan/models/utils.py]: oh, it seems the moe part has `top_k` and `num_experts` in `model_args ` as well. Yeah NVM we can keep `model_args` as input arg for now.

Comment (wwwjn) [reply torchtitan/models/utils.py]: one minor thing is that we might need `n_layers` , and  `n_heads`, and `head_dim`, and `dim` as the function args. This is because in Qwen3, the `head_dim != dim / n_heads`

Comment (shuhuayu) [reply torchtitan/models/utils.py]: Thanks for the suggestions! Code updated accordingly.

File: torchtitan/models/utils.py [hunk 12]
```diff
@@ -361,28 +361,43 @@ def _concatenate_expert_weights(
 
 
 def get_dense_model_nparams_and_flops(
-    model_args: BaseModelArgs, model: nn.Module, seq_len: int
+    model_args: BaseModelArgs,
+    model: nn.Module,
+    n_layers: int,
+    n_heads: int,
```
Comment (tianyu-l) [inline torchtitan/models/utils.py]: oh sorry I didn't make myself clear -- I think if we need to pass in `model_args` anyway, we shouldn't pass in n_layers, n_heads redundantly.
